create GlobalLinterMessages component #872

mirefly opened the pull request:
Fixes #872 
@kumar303 
In refactoring, I found scrolling to linter messages in different files is not work well (Not sure why..) https://github.com/mozilla/addons-code-manager/issues/887#issuecomment-506862273. 
I tested some pages, this refactoring does not change behaviors when navigating between linter messages.

kumar303 commented on src/components/GlobalLinterMessages/index.tsx:
```diff
@@ -0,0 +1,33 @@
+const GlobalLinterMessages = ({
+  className,
+  children,
+  scrollToSelectedLine,
```
GlobalLinterMessages doesn't need to know about *what* the ref is used for. It's just a ref. Thus, the name of this prop should be something generic like `ref` or `containerRef`.

kumar303 commented on src/components/CodeView/index.tsx:
```diff
@@ -14,6 +14,7 @@ import refractor from '../../refractor';
 import { getLanguageFromMimeType } from '../../utils';
 import { Version } from '../../reducers/versions';
 import LinterProvider, { LinterProviderInfo } from '../LinterProvider';
+import GloablLinterMessages from '../GlobalLinterMessages';
```
There's a typo in this import.

kumar303 commented on src/components/CodeView/index.tsx:
```diff
@@ -68,17 +69,24 @@ export class CodeViewBase extends React.Component<Props> {
+              : undefined
+          }
+        >
+          {globalLinterMessages}
```
Rather than make the caller create each `<LinterMessage>` component, it would save some duplication to accept a list of messages, like:

```jsx
<GlobalLinterMessages messages={selectedMessageMap.global} />
```

With that, `GlobalLinterMessages` can be responsible for mapping each array item to `<LinterMessage>`.

kumar303 commented on src/components/GlobalLinterMessages/index.tsx:
```diff
@@ -0,0 +1,33 @@
+import { scrollToSelectedLine as _scrollToSelectedLine } from '../CodeView';
+
+type PublicProps = {
+  className?: string;
```
The spec needs a test to make sure that a `className` prop results in an actual class added to the element. Here's an example:

https://github.com/mozilla/addons-code-manager/blob/365b8958bd517c444f4af1008a0eb3beb211cb4b/src/components/FullscreenGrid/ContentShell.spec.tsx#L23

kumar303 commented on src/components/GlobalLinterMessages/index.tsx:
```diff
@@ -0,0 +1,33 @@
+type PublicProps = {
+  className?: string;
+  children: React.ReactNode[];
+  scrollToSelectedLine?: typeof _scrollToSelectedLine;
```
The spec needs a test to make sure that the ref prop gets added to the container. Unfortunately, you'll need to use `mount()` for that (but just for that one test). Here's an example:

https://github.com/mozilla/addons-code-manager/blob/da479dba41330c2714f4d9a722c3f29945dd8f1a/src/components/CodeView/index.spec.tsx#L249-L256

kumar303 commented on src/components/GlobalLinterMessages/index.tsx:
```diff
@@ -0,0 +1,33 @@
+  scrollToSelectedLine,
+}: PublicProps) => {
+  return (
+    <React.Fragment>
```
You can do this without a fragment. It's fine to return `null` in a component.

mirefly commented:
@kumar303 I believe I addressed all comments.
I have one question about the type of `containerRef`. Now I used the type of `scrollToSelectedLine` in `CodeView`, but not sure why to use `HTMLTableRowElement` instead of `HTMLElement`?
https://github.com/mozilla/addons-code-manager/blob/2fe6782291e8b13f0a0ee05ca2ba863c27ca6162/src/components/CodeView/index.tsx#L40

kumar303 commented on src/components/GlobalLinterMessages/index.spec.tsx:
```diff
@@ -0,0 +1,63 @@
+
+  it('renders multiple linter messages', () => {
+    const messages = [
+      createInternalMessage(fakeExternalLinterMessage),
```
The `uid` is used as a key, so it's a good idea to make them unique:

```diff
diff --git a/src/components/GlobalLinterMessages/index.spec.tsx b/src/components/GlobalLinterMessages/index.spec.tsx
index 098eb24..a241588 100644
--- a/src/components/GlobalLinterMessages/index.spec.tsx
+++ b/src/components/GlobalLinterMessages/index.spec.tsx
@@ -32,8 +32,14 @@ describe(__filename, () => {
 
   it('renders multiple linter messages', () => {
     const messages = [
-      createInternalMessage(fakeExternalLinterMessage),
-      createInternalMessage(fakeExternalLinterMessage),
+      createInternalMessage({
+        ...fakeExternalLinterMessage,
+        uid: 'first-message',
+      }),
+      createInternalMessage({
+        ...fakeExternalLinterMessage,
+        uid: 'second-message',
+      }),
     ];
     const root = render({ messages });
     const div = root.find('div');

```

kumar303 commented on src/components/GlobalLinterMessages/index.spec.tsx:
```diff
@@ -0,0 +1,63 @@
+    const root = render({ messages: [] });
+    const div = root.find('div');
+
+    expect(div).toHaveLength(0);
```
If the top-level wrapper changes from a `div` to another component (for example), this test shouldn't fail. Here's a more robust way to make the same assertion:

```diff
diff --git a/src/components/GlobalLinterMessages/index.spec.tsx b/src/components/GlobalLinterMessages/index.spec.tsx
index 098eb24..74215f6 100644
--- a/src/components/GlobalLinterMessages/index.spec.tsx
+++ b/src/components/GlobalLinterMessages/index.spec.tsx
@@ -25,9 +25,8 @@ describe(__filename, () => {
 
   it('renders no linter messages', () => {
     const root = render({ messages: [] });
-    const div = root.find('div');
 
-    expect(div).toHaveLength(0);
+    expect(root.type()).toEqual(null);
   });
 
   it('renders multiple linter messages', () => {

```

mirefly commented:
Thanks for reviewing @kumar303 . I updated and rebased it.

kumar303 commented on src/components/CodeView/index.tsx:
```diff
@@ -36,7 +37,9 @@ const isLineSelected = (
 
-export const scrollToSelectedLine = (element: HTMLTableRowElement | null) => {
+export const scrollToSelectedLine = (
+  element: HTMLTableRowElement | HTMLDivElement | null,
```
👍

kumar303 commented on src/components/GlobalLinterMessages/index.spec.tsx:
```diff
@@ -0,0 +1,66 @@
+  };
+
+  it('renders no linter messages', () => {
+    const root = render({ messages: [] });
```
This needs one more test: `render({ messages: undefined })`. That's an important difference because an empty array and an undefined variable will be handled differently.

kumar303 commented on src/components/GlobalLinterMessages/index.tsx:
```diff
@@ -0,0 +1,27 @@
+
+export type PublicProps = {
+  className?: string;
+  messages?: LinterMessageType[] | null | void;
```
This property should not be optional, i.e. you can remove the question mark. This is because the caller would be mistaken to omit the property, like `<GlobalLinterMessages />`. We want TypeScript to tell them that they forgot to specify `messages`.

kumar303 commented:
The snyk failure has been fixed on master so you can rebase again to get the fix (sorry about all the rebasing!)

mirefly commented:
Updated!  :) @kumar303
